internal/incident/incidents.go

```diff
@@ -25,38 +27,122 @@ var (
 )
 
 // LoadOpenIncidents loads all active (not yet closed) incidents from the database and restores all their states.
-// Returns error ony database failure.
+// Returns error on any database failure.
 func LoadOpenIncidents(ctx context.Context, db *icingadb.DB, logger *logging.Logger, runtimeConfig *config.RuntimeConfig) error {
 	logger.Info("Loading all active incidents from database")
 
-	var objectIDs []types.Binary
-	err := db.SelectContext(ctx, &objectIDs, `SELECT object_id FROM incident WHERE "recovered_at" IS NULL`)
-	if err != nil {
-		logger.Errorw("Failed to load active incidents from database", zap.Error(err))
+	g, chCtx := errgroup.WithContext(ctx)
 
-		return errors.New("failed to fetch open incidents")
-	}
+	incidents := make(chan *Incident)
+	g.Go(func() error {
+		defer close(incidents)
 
-	for _, objectID := range objectIDs {
-		obj, err := object.LoadFromDB(ctx, db, objectID)
+		rows, err := db.QueryxContext(chCtx, db.BuildSelectStmt(new(Incident), new(Incident))+` WHERE "recovered_at" IS NULL`)
 		if err != nil {
-			logger.Errorw("Failed to retrieve incident object from database", zap.Error(err))
-			continue
+			return err
 		}
+		// In case the incidents in the loop below are successfully traversed, rows is automatically closed and an
+		// error is returned (if any), making this rows#Close() call a no-op. Escaping from this function unexpectedly
+		// means we have a more serious problem, so in either case just disregard the error here.
+		defer func() { _ = rows.Close() }()
+
+		for rows.Next() {
+			i := NewIncident(db, nil, runtimeConfig, nil)
+			if err := rows.StructScan(i); err != nil {
+				return err
+			}
 
-		incident, _, err := GetCurrent(ctx, db, obj, logger, runtimeConfig, false)
-		if err != nil {
-			continue
+			select {
+			case incidents <- i:
+			case <-chCtx.Done():
+				return chCtx.Err()
+			}
 		}
 
-		incident.RetriggerEscalations(&event.Event{
-			Time:    time.Now(),
-			Type:    event.TypeInternal,
-			Message: "Incident reevaluation at daemon startup",
-		})
-	}
+		return rows.Err()
+	})
+
+	g.Go(func() error {
+		bulks := com.Bulk(chCtx, incidents, db.Options.MaxPlaceholdersPerStatement, com.NeverSplit[*Incident])
+
+		for {
+			select {
+			case <-chCtx.Done():
+				return chCtx.Err()
+			case bulk, ok := <-bulks:
+				if !ok {
+					return nil
+				}
+
+				g.Go(func() error {
+					chunkLen := len(bulk)
+					objectIds := make([]types.Binary, chunkLen)
+					incidentIds := make([]int64, chunkLen)
+					incidentsById := make(map[int64]*Incident, chunkLen)
+					incidentsByObjId := make(map[string]*Incident, chunkLen)
+
+					for _, i := range bulk {
+						incidentsById[i.ID()] = i
+						incidentsByObjId[i.ObjectID.String()] = i
+
+						objectIds = append(objectIds, i.ObjectID)
+						incidentIds = append(incidentIds, i.ID())
```

Comment:
You're initializing the slices to that `len(objectIds) = len(incidentIds) = len(bulk)` and then append to the first two. I.e. those will end up having `len(objectIds) = len(incidentIds) = 2 * len(bulk)` in the end where the first half is filled with zero values.

Reply:
🤦‍♂️ !